packages/cli/generators/example/clone-example.js

```diff
@@ -0,0 +1,107 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/cli
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+'use strict';
+
+const promisify = require('../../lib/promisify');
+
+const yauzl = require('yauzl');
+const fs = require('fs');
+const mkdirp = require('mkdirp');
+const os = require('os');
+const path = require('path');
+const request = require('request');
+const rimraf = promisify(require('rimraf'));
+
+const GITHUB_ZIP_URL =
+  'https://github.com/strongloop/loopback-next/archive/master.zip';
+
+module.exports = function cloneExampleFromGitHub(
+  exampleName,
+  log = console.log.bind(console)
+) {
+  const outDir = `loopback4-example-${exampleName}`;
+
+  let tempDir, zipPath;
+
+  return (
+    createTempDir()
+      .then(tmp => {
+        tempDir = tmp;
+        zipPath = path.join(tmp, 'loopback4-master.zip');
+        return download(GITHUB_ZIP_URL, zipPath, log);
+      })
+      .then(() => extractExample(zipPath, exampleName, outDir, log))
+      .then(() => {
+        return outDir;
+      })
+      // Poor man's finally handler
+      .then(
+        success => rimraf(tempDir).then(() => success),
+        err => rimraf(tempDir).then(() => Promise.reject(err))
+      )
+  );
+};
+
+const mkdtemp = promisify(fs.mkdtemp);
+function createTempDir() {
+  const prefix = path.join(os.tmpdir(), 'lb-');
+  return mkdtemp(prefix);
+}
+
+function download(url, filePath, log) {
+  log(`Downloading ${GITHUB_ZIP_URL}`);
+  return new Promise((resolve, reject) => {
+    request(url)
+      .pipe(fs.createWriteStream(filePath))
+      .on('error', reject)
+      .on('finish', () => resolve());
+  });
+}
+
+function extractExample(zipPath, exampleName, outDir, log) {
+  const sourceDir = `loopback-next-master/packages/example-${exampleName}`;
+
+  log(`Extracting ${exampleName} to ${outDir}`);
+
+  return new Promise((resolve, reject) => {
+    yauzl.open(zipPath, {lazyEntries: true}, function(err, zipFile) {
+      if (err) return reject(err);
+
+      zipFile.readEntry();
+      zipFile.on('entry', entry =>
+        extractEntry(entry, zipFile, err => {
+          if (err) {
+            reject(err);
+            zipFile.close();
+          }
+          zipFile.readEntry();
+        })
+      );
+      zipFile.on('close', () => resolve());
+    });
+
+    function extractEntry(entry, zipFile, next) {
+      // Skip optional directory entries
+      if (entry.fileName.endsWith('/')) return next();
```

Comment:
`path.sep` constant should be used for windows compatibility.

Reply:
IIRC, zip entry file names always use `/`'.

Reply:
Good catch. According to yauzl's documentation, directory entries in ZIP end with `/` character - see https://www.npmjs.com/package/yauzl#usage:

```js
    if (/\/$/.test(entry.fileName)) {
      // Directory file names end with '/'.
      // Note that entires for directories themselves are optional.
      // An entry's fileName implicitly requires its parent directories to exist.
      zipfile.readEntry();
    }
```

See also [ZIP format specification](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT), section 4.4.17.1:

> The name of the file, with optional relative path.
> The path stored MUST not contain a drive or
> device letter, or a leading slash.  All slashes
> MUST be forward slashes '/' as opposed to
> backwards slashes '\' for compatibility with Amiga
> and UNIX file systems etc.  If input came from standard
 > input, there is no file name field.